### PR TITLE
fix(compiler): missing environment variable runtime error

### LINKED ISSUE #1966: Reference to `this` is unknown inside cloud.Api method definition
### Community Note

> Please vote by adding a 👍 reaction to the issue to help us prioritize.
> If you are interested to work on this issue, please leave a comment.

### I tried this:

`wing it`
```ts
bring cloud;

class TaskList {  
  init() {
  }
    
  inflight add(title: str): str {
    let id = "1";
    return id;
  }
}
      
class TaskListApi {
  api: cloud.Api;
  task_list: TaskList;
                
  init(task_list: TaskList) {
    this.task_list = task_list;
    this.api = new cloud.Api();
        
    this.api.post("/tasks", inflight (req: cloud.ApiRequest): cloud.ApiResponse => {
      let id = this.task_list.add("title");
      return cloud.ApiResponse { status:201, body: id };
    });
  }
}

let task_list = new TaskList();
let t = new TaskListApi(task_list);
```
Then in wing console I tried to `POST` to `/tasks` a simple request with `body: {"title": "Yossarian"}` and `content-type` header `application/json`.

### I expected this:

Wing console to print: success.

### Instead, this happened:

```sh
evalmachine.<anonymous>:19
  async handle(req) { const { this } = this; {
                              ^^^^

SyntaxError: Unexpected token 'this'
    at new Script (node:vm:100:7)
    at createScript (node:vm:265:10)
    at Object.runInContext (node:vm:296:10)
    at /Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:238:149
    at new Promise (<anonymous>)
    at n2t (/Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:237:9364)
    at Object.activity (/Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:237:8767)
    at Object.withTrace (/Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:238:8269)
    at XH.invoke (/Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:237:8689)
    at /Applications/Wing Console.app/Contents/Resources/app.asar/dist/vite/electron/main/index.js:237:2531
```

### Is there a workaround?

Yes, this seems to work (notice assignment of `self`):
```ts
bring cloud;

class TaskList {  
  init() {
  }
    
  inflight add(title: str): str {
    let id = "1";
    return id;
  }
}
      
class TaskListApi {
  api: cloud.Api;
  task_list: TaskList;
                
  init(task_list: TaskList) {
    this.task_list = task_list;
    this.api = new cloud.Api();

    let self = this;
        
    this.api.post("/tasks", inflight (req: cloud.ApiRequest): cloud.ApiResponse => {
      let id = self.task_list.add("title");
      return cloud.ApiResponse { status:201, body: id };
    });
  }
}

let task_list = new TaskList();
let t = new TaskListApi(task_list);
```
This code now runs into a different issue #1969

### Component:

Compiler

### Wing Version:

0.8.26

### Node.js Version:

19.3.0

### Platform(s):

MacOS

### Anything else?

_No response_

### LINKED ISSUE #2082: Resource inflight client created within init() error
### I tried this

`wing test example.w`
```js
bring cloud;

resource Sleeper {
  init() {}
  extern "./sleeper.js" inflight sleep(ms: num): str;
}

resource MyProcessor {
  b: cloud.Bucket;
  q: cloud.Queue;

  init() {
    this.q = new cloud.Queue();
    this.b = new cloud.Bucket();

    this.q.add_consumer(inflight (message: str) => {
      log("processing ${message}..");
      this.b.put("file.txt", message);
    });
  }

  inflight publish(s: str) {
    this.q.push(s);
  }

  inflight assertion(): bool {
    return this.b.get("file.txt") == "cool";
  }
}

let p = new MyProcessor();
let s = new Sleeper();

new cloud.Function(inflight () => {
	p.publish("cool");
	s.sleep(200); // give queue enough time		
	assert(p.assertion());
}) as "test";
```

Extern code:
```js
exports.sleep = function(ms) {
  console.log('sleeping for ' + ms + ' ms');
  return new Promise(resolve => setTimeout(resolve, ms));
}
```

### I expected this:

Test to pass

### Instead, this happened

No object was written to bucket. 
image

### Is there a workaround?

Not sure

### Component

Compiler, SDK

### Wing Version

0.11.1

### Wing Console Version

_No response_

### Node.js Version

_No response_

### Platform(s)

_No response_

### Anything else?

Further investigation into the sdk revealed there is an error invoking the inflight function on the queue consumer
```
 ﬌ mywing test examples/tests/valid/hmm.w
✔ Compiling to sim...
/Users/hasan/repos/monada/wing/examples/tests/valid/target/test/hmm.wsim/.wing/cloud.queue-addconsumer-0088483a_c8fe5ad5.js:10
  const { this } = this;
          ^

SyntaxError: Unexpected keyword 'this'
    at makeNiceSyntaxError (/Users/hasan/repos/monada/wing/libs/wingsdk/node_modules/vm2/lib/transformer.js:41:16)
    at transformer (/Users/hasan/repos/monada/wing/libs/wingsdk/node_modules/vm2/lib/transformer.js:84:8)
    at NodeVM.run (/Users/hasan/repos/monada/wing/libs/wingsdk/node_modules/vm2/lib/nodevm.js:414:17)
    at NodeVM.runFile (/Users/hasan/repos/monada/wing/libs/wingsdk/node_modules/vm2/lib/vm.js:534:15)
```

Looking into the generated client
image

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

### PR DESCRIPTION
**Pull request by Chriscbr**
Fixes #1966
Fixes #2082

Addresses a bug in the compiler that resulted in "missing environment variable" errors appearing in some situations where inflight functions were created inside of class constructors.

Take the following example:

```js
class MyProcessor {
  b: cloud.Bucket;
  q: cloud.Queue;

  init() {
    this.q = new cloud.Queue();
    this.b = new cloud.Bucket();

    this.q.addConsumer(inflight (message: str) => {
      log("processing ${message}..");
      this.b.put("file.txt", message);
    });
  }
}
```

`this.q.addConsumer` creates a `cloud.Function` resource to run the inflight code. That cloud resource was originally receiving the minimal permissions needed to call the "put" method on "this.b". This caused an issue because in the current way the compiler generates inflight code, it is bundling and instantiating the entire client of "this" - i.e. the inflight client of MyProcessor.

This causes issues because the inflight client code of MyProcessor is a JavaScript class that looks like this:

```js
  class MyProcessor {
    constructor({ b, q }) {
      this.b = b;
      this.q = q;
    }
  }
```

... and declaring `q` requires instantiating a `cloud.Queue` client, which requires these extra environment variables.

The current fix I've added is to grant "inflight init" permissions to the entirety of `this` despite the fact that only the child (`this.b`) is used.

In the future we might want to generate inflight code differently so only `this.b` has to be bundled, but I think it might require more work.

- [x] PR title matches [Winglang's style guide](https://docs.winglang.io/contributors/pull_requests#how-are-pull-request-titles-formatted)
- [x] PR description explains motivation and solution
- [x] Tests added
- [x] Docs updated
- [ ] Needs end-to-end tests (`pr/e2e-full` label)

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: tools/hangar/__snapshots__/test_corpus/api.w_compile_tf-aws.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ module.exports = function({ counter }) {
     }
     async handle(request)  {
       {
-        const count = (await counter.inc());
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/test_corpus/api.w_compile_tf-aws.md:11]: Because we added "skip" to this test, the snapshot wasn't updated for around a month, so there's a lot of unrelated changes here

**Review by eladb — APPROVED** [review]
Ship it!!
